### PR TITLE
Query Explanation

### PR DESCRIPTION
**Pull request by PradyumnaKrishna**
* [ ] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes
- A param `explain` is added to QueryAPI, if true then explanation returned by the `Explain()` method of the query having structure `ExplainOutputNode` is returned in response.
- Query Explanation is added under new field in response that is `thanosInfo`.
- Adds explanation to thanos UI.
- Disables `explain` checkbox on `prometheus` engine selection.
- Makes explanation UI better, by adding alert background and horizontal scrolling.

**Screenshots**
Disabled button
Explanation

## Verification

Tested locally

### CODE REVIEW

File: pkg/ui/react-app/src/pages/graph/Panel.tsx [hunk 1]
```diff
@@ -56,6 +71,8 @@ export interface PanelOptions {
   usePartialResponse: boolean;
   storeMatches: Store[];
   engine: string;
+  explain: boolean;
+  disableCheckbox: boolean;
```
Comment (GiedriusS) [inline pkg/ui/react-app/src/pages/graph/Panel.tsx:75]: Let's change it to `disableExplainCheckbox` to make it crystal clear we're talking about the "explain" checkbox.

**Review by GiedriusS — COMMENTED** [review]
:muscle:

File: pkg/ui/react-app/src/components/ListTree.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      const item = (
+        <React.Fragment>
+          <ListGroupItem
+            className={`bg-transparent p-0 border-0 ${parentId ? `rounded-0 ${lvl ? 'border-bottom-0' : ''}` : ''}`}
+          >
+            {
+              <div className={`d-flex align-items-center`} style={{ paddingLeft: `${25 * lvl}px` }}>
+                {node.children && (
+                  <div className="pl-0 btn text-primary" style={{ cursor: 'inherit' }} color="link">
+                    {state[id] ? '-' : '+'}
```
Comment (saswatamcode) [inline pkg/ui/react-app/src/components/ListTree.tsx:39]: This is already really good! I wonder tho if we can use some unicode symbol instead to better represent this?

File: pkg/ui/react-app/src/components/ListTree.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  const [state, setState] = useState<mapping>({});
+  const toggle = (e: React.MouseEvent<HTMLDivElement>) => {
+    const el = e.target as HTMLDivElement;
+    const id = el.getAttribute('id');
+    if (id) {
+      setState({ ...state, [id]: !state[id] });
+    }
+  };
+
+  const mapper = (nodes: NodeTree[], parentId?: any, lvl?: any) => {
```
Comment (saswatamcode) [inline pkg/ui/react-app/src/components/ListTree.tsx:27]: Maybe we can add a comment on what `mapper` is doing recursively? 🙂

File: pkg/ui/react-app/src/components/ListTree.tsx [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+interface NodeProps extends InputProps {
+  node: NodeTree | null;
+}
+
+const ListTree: React.FC<NodeProps> = ({ id, node }) => {
+  type mapping = {
+    [key: string]: boolean;
+  };
+
+  const [state, setState] = useState<mapping>({});
```
Comment (saswatamcode) [inline pkg/ui/react-app/src/components/ListTree.tsx:18]: Let's make naming a bit more descriptive here!
```suggestion
  const [mapping, setMapping] = useState<mapping>({});
```

File: pkg/ui/react-app/src/components/ListTree.tsx [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    const el = e.target as HTMLDivElement;
+    const id = el.getAttribute('id');
+    if (id) {
+      setState({ ...state, [id]: !state[id] });
+    }
+  };
+
+  const mapper = (nodes: NodeTree[], parentId?: any, lvl?: any) => {
+    return nodes.map((node: NodeTree, index: number) => {
+      const id = `${index}-${parentId ? parentId : 'top'}`.replace(/[^a-zA-Z0-9-_]/g, '');
```
Comment (saswatamcode) [inline pkg/ui/react-app/src/components/ListTree.tsx:29]: Why are we constructing the id this way?

**Review by saswatamcode — APPROVED** [review]
Amazing job! This already looks good! 💪🏻 

A few small nits/questions. 🙂

**Review by GiedriusS — COMMENTED** [review]
If I just enable "Explain", it shows me an empty, blue rectangle. Perhaps we could only show it after the first query?
Screenshot from 2023-05-19 11-08-42

File: pkg/ui/react-app/src/pages/graph/Panel.tsx [hunk 6]
```diff
@@ -348,44 +389,64 @@ class Panel extends Component<PanelProps & PathPrefixProps, PanelState> {
...
+                onChange={this.handleChangeExplain}
+                checked={options.explain}
+                disabled={options.disableExplainCheckbox}
+              >
+                Explain
+              </Checkbox>
+            </div>
+          </Col>
+        </Row>
+        <Row hidden={!options.explain}>
```
Comment (GiedriusS) [inline pkg/ui/react-app/src/pages/graph/Panel.tsx:445]: Probably here we need to add `&& this.state.data`

File: pkg/ui/react-app/src/components/ListTree.tsx (hunk 5 above)
Comment (PradyumnaKrishna) [reply pkg/ui/react-app/src/components/ListTree.tsx:29]: I don't know much, I found it online and changed it a little.

**Review by GiedriusS — APPROVED** [review]
LGTM :medal_military:

**Review by saswatamcode — APPROVED** [review]
LGTM! Thanks for awesome work!
Let's merge and iterate on this.

**Pull request merged**